frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts

```diff
@@ -0,0 +1,681 @@
+    AimOutlined,
+    CheckOutlined,
+    SyncOutlined,
+    TagOutlined,
+    ClockCircleOutlined,
+    UserOutlined,
+    UsergroupAddOutlined,
+    ExperimentOutlined,
+    SettingOutlined,
+    MessageOutlined,
+    TeamOutlined,
+    LinkOutlined,
+    CalculatorOutlined,
+    FunnelPlotOutlined,
+    GatewayOutlined,
+    InteractionOutlined,
+    MailOutlined,
+    KeyOutlined,
+    VideoCameraOutlined,
+    SendOutlined,
+    LogoutOutlined,
+    PlusOutlined,
+    LineChartOutlined,
+} from '@ant-design/icons'
+import { DashboardType } from '~/types'
+import api from 'lib/api'
+import { appUrlsLogic } from '../AppEditorLink/appUrlsLogic'
+import { copyToClipboard, isURL } from 'lib/utils'
+import { personalAPIKeysLogic } from '../PersonalAPIKeys/personalAPIKeysLogic'
+
+// If CommandExecutor returns CommandFlow, flow will be entered
+export type CommandExecutor = () => CommandFlow | void
+
+export interface CommandResultTemplate {
+    icon: any // any, because Ant Design icons are some weird ForwardRefExoticComponent type
+    display: string
+    synonyms?: string[]
+    prefixApplied?: string
+    executor?: CommandExecutor | true // true means "just clear input"
+    guarantee?: boolean // show result always and first, regardless of fuzzy search
+}
+
+export type CommandResult = CommandResultTemplate & {
+    source: Command | CommandFlow
+    index?: number
+}
+
+export type CommandResolver = (
+    argument?: string,
+    prefixApplied?: string
+) => CommandResultTemplate[] | CommandResultTemplate | null
+
+export interface Command {
+    key: string // Unique command identification key
+    prefixes?: string[] // Command prefixes, e.g. "go to". Prefix-less case is dynamic base command (e.g. Dashboard)
+    resolver: CommandResolver | CommandResultTemplate[] | CommandResultTemplate // Resolver based on arguments (prefix excluded)
+    scope: string
+}
+
+export interface CommandFlow {
+    icon?: any
+    instruction?: string
+    resolver: CommandResolver | CommandResultTemplate[] | CommandResultTemplate
+    scope: string
+}
+
+export type CommandRegistrations = {
+    [commandKey: string]: Command
+}
+
+export type RegExpCommandPairs = [RegExp | null, Command][]
+
+const RESULTS_MAX = 5
+
+const GLOBAL_COMMAND_SCOPE = 'global'
+
+function resolveCommand(source: Command | CommandFlow, argument?: string, prefixApplied?: string): CommandResult[] {
+    // run resolver or use ready-made results
+    let results = source.resolver instanceof Function ? source.resolver(argument, prefixApplied) : source.resolver
+    if (!results) return [] // skip if no result
+    if (!Array.isArray(results)) results = [results] // work with a single result and with an array of results
+    const resultsWithCommand: CommandResult[] = results.map((result) => {
+        return { ...result, source }
+    })
+    return resultsWithCommand
+}
+
+export const commandPaletteLogic = kea<commandPaletteLogicType<Command, CommandRegistrations>>({
```

Comment:
This should be changed to `commandPaletteLogicType<Command, CommandRegistrations, CommandResult, CommandFlow, RegExpCommandPairs>`